src/languageclient.rs

```diff
@@ -395,47 +395,105 @@ impl State {
-                None,
-                "nvim_buf_add_highlight",
-                json!([0, 0, "Error", 1, 1, 1]),
-            )?);
-            self.update(|state| {
-                state.highlight_source = source;
-                Ok(())
-            })?;
-        }
-        let source = source.ok_or_else(|| err_msg("Empty highlight source id"))?;
         let diagnosticsDisplay = self.get(|state| Ok(state.diagnosticsDisplay.clone()))?;
 
-        // TODO: Optimize.
-        self.call::<_, Option<u8>>(None, "nvim_buf_clear_highlight", json!([0, source, 1, -1]))?;
-        for dn in diagnostics {
-            let severity = dn.severity.unwrap_or(DiagnosticSeverity::Information);
-            let hl_group = diagnosticsDisplay
-                .get(&severity.to_int()?)
-                .ok_or_else(|| err_msg("Failed to get display"))?
-                .texthl
-                .clone();
-            self.call::<_, u8>(
+        // Highlight.
+        if self.get(|state| Ok(state.is_nvim))? {
+            let mut source: Option<u64> = self.get(|state| Ok(state.highlight_source))?;
+            if source.is_none() {
+                source = Some(self.call(
+                    None,
+                    "nvim_buf_add_highlight",
+                    json!([0, 0, "Error", 1, 1, 1]),
+                )?);
+                self.update(|state| {
+                    state.highlight_source = source;
+                    Ok(())
+                })?;
+            }
+            let source = source.ok_or_else(|| err_msg("Empty highlight source id"))?;
+
+            // TODO: Optimize.
+            self.call::<_, Option<u8>>(
                 None,
-                "nvim_buf_add_highlight",
-                json!([
-                    0,
-                    source,
-                    hl_group,
-                    dn.range.start.line,
-                    dn.range.start.character,
-                    dn.range.end.character,
-                ]),
+                "nvim_buf_clear_highlight",
+                json!([0, source, 1, -1]),
             )?;
+            for dn in diagnostics {
+                let severity = dn.severity.unwrap_or(DiagnosticSeverity::Information);
+                let hl_group = diagnosticsDisplay
+                    .get(&severity.to_int()?)
+                    .ok_or_else(|| err_msg("Failed to get display"))?
+                    .texthl
+                    .clone();
+
+                self.call::<_, u8>(
+                    None,
+                    "nvim_buf_add_highlight",
+                    json!([
+                        0,
+                        source,
+                        hl_group,
+                        dn.range.start.line,
+                        dn.range.start.character,
+                        dn.range.end.character,
+                    ]),
+                )?;
+            }
+        } else {
+            self.call::<_, u8>(None, "clearmatches", json!([]))?;
```

Comment:
`matchaddpos` returns an id that you can call `matchdelete` with to clear only that match. Would it be worthwhile to look at storing these ids in between calls to prevent conflicts?

Reply:
It's up to you if you want to include that within this PR.

I'm OK with current implementation.

Reply:
From doing a quick GitHub code search, I couldn't find any major plugins using `matchaddpos` but I've updated it to be a good citizen anyway